Configuration table for mesh target panel nodes

ferhoyos opened the pull request:
### Describe the change

This PR presents mesh node configuration in a table format instead of plain JSON, adding copy and download buttons.

![image](https://github.com/kiali/kiali/assets/122779323/38efdbe7-827a-45d0-a3e0-3f392b08e84a)

**Bonus**:
- Update `json-beautify` library to add Typescript types.
- Add typescript types for `react-copy-to-clipboard` library.
- Refactor i18n labels that use variables with the interpolation feature.
- Refactor `DebugInformation` to use the shared `JsonTable` component.
- Reorganize some components for better classificiation (all components belonging to one folder):
  - Create a `Table` folder to group `SimpleTable` and `JsonTable`.
  - Create a `Dropdown` folder to group all dropdown components.
  - Create a `Select` folder and move the `SimpleSelect` component there.
  - Move `PfSpinner` to the `Pf` folder.
  - Move `TextOfLink` to the `Link` folder.
- Move the `download` function to `utils` as it is used by at least 3 different components.
- Refactor `IstioConfigPreview` since it contained some `trafficPolicies` hardcodes and it should be a generic component. Modify the copy tooltip to `Copied` when the user clicks to copy button to indicate the user that the text has been successfully copied to the clipboard.
- Convert `GraphSecondaryMasthead` component to React hook
- Fix `HealthIndicator` unit test
- Lint fixes

### Steps to test the PR

1. Go to mesh page
2. Click on any mesh node (e.g., Jaeger, Grafana, Prometheus, etc.)
3. Check that the mesh node configuration is displayed in a table format
4. Check that the `Copy` button copies the configuration json in the clipboard
5. Check that the `Download` button downloads the configuration json file
6. For the bonus, check that the DebugInformation in the toolbar and the Istio config object preview work as expected (only refactor work)

![image](https://github.com/kiali/kiali/assets/122779323/29beead6-a352-48b8-a2f5-10829cc69bed)

![image](https://github.com/kiali/kiali/assets/122779323/c9306921-3ca3-49b3-94cd-72815f37472e)

### Automation testing

N/A

### Issue reference

Closes #7379

hhovsepy reviewed (dismissed):
Works fine for me:
![Screenshot from 2024-06-10 09-29-35](https://github.com/kiali/kiali/assets/604313/82034eb8-34a9-49d0-b9dd-8e44218504b0)

Also config yaml and debug info are fine.

jshaughn commented on frontend/src/types/Health.ts:
```diff
@@ -19,11 +19,17 @@ interface HealthConfig {
 }
 
+export const enum HealthItemType {
+  TRAFFIC_STATUS = 'traffic_status',
+  POD_STATUS = 'pod_status'
```
Could be alphabetized.

jshaughn reviewed (approved):
The minor code-reorg looks good and I see your usual improvements around class->FC, i18n, typing, px>rem, etc :-)   The new JsonTable component is nice. 
TargetPanelConfigTable seems to be working fine.   Having Copy/Download everywhere is great!  Looks good!

nrfox commented:
Is it possible to render the configuration as yaml? It could be more readable especially for deeply nested fields and it would also match the format that your istio and kiali configs are in: `kubectl get configmaps -n istio-system istio -o yaml`

ferhoyos commented:
@nrfox yes, it is quite straightforward to render the configuration as yaml.

Kiali configuration
- JSON:
![image](https://github.com/kiali/kiali/assets/122779323/a7d484e6-9380-411c-b594-60ebe6cf302d)

- YAML:
![image](https://github.com/kiali/kiali/assets/122779323/dca971c3-3ef2-4dc7-8ce3-6e235fabebf3)

Debug information:
- JSON:
![image](https://github.com/kiali/kiali/assets/122779323/75750ca6-cbb8-499f-8dc9-28a847fd1daf)

- YAML:
![image](https://github.com/kiali/kiali/assets/122779323/fcb36397-ec47-413b-ad3f-749c2c7d5176)

Additional State:
- JSON:
![image](https://github.com/kiali/kiali/assets/122779323/c9500045-437b-4918-9e25-df1c75f4cbfa)

- YAML:
![image](https://github.com/kiali/kiali/assets/122779323/a1097aec-48ab-47fa-bf68-fd5f69f071b0)

The YAML format seems to offer better readability in tables than JSON. Regarding additional state, IMO, it is quite similar. However, in the Kubernetes world, YAML is preferred over JSON.

What do you think @jshaughn @hhovsepy?

hhovsepy commented:
IMO this new table format is more readable, additionally there can be `Download as JSON` and `Download as YAML` both buttons.

nrfox commented:
Sorry I'm coming in late looking at this so I don't have much context but why have the table at all? Could you just use something like the ace editor to show the yaml unbroken? With the table the top level keys are all separated which imo makes it harder to read. The table is great for labels and annotations which are always going to be `map[string]string`. You can have annotations that are json strings but that's more for one-off configuration and it's still just a string. The kiali and istio configs are always going to be nested data.

Also do we want to filter out empty values? There's a lot of config values whose value is just `{}` or `""` which makes the config more noisy.

ferhoyos commented:
No problem @nrfox, you are always welcome to the party :wink: 

IMO I would use ace-editor for editable yamls or for complex objects like `Additional State`. Kiali and Istio config are quite simple, mainly `map[string]string` except for some fields, and the table format is a more beautiful way to display this information instead of plain yaml or json:

![image](https://github.com/kiali/kiali/assets/122779323/d11f1540-d689-4cdf-a26b-853c173a1f3e)

If the top level keys are very separated we can play with the table column widths...

Regarding the empty values... maybe it is good the user knows that this config value is empty. I'm not sure here.

@hhovsepy I would keep it simple and download only in one format (e.g., yaml). The user can always use yaml<->json conversion tool (as said Kiali and Istio configs are quite simple objects).

jshaughn commented:
@ferhoyos , I like the YAML suggestion, looks better.  @nrfox , I think the table approach was just to match the approach previously taken by our `Debug Info` presentation.  I don't know whether that was UX-influenced or our own doing.  My suggestion is to keep the new table format, I think it looks good in most cases.  But, make the switch to YAML.  For simplicity, I'd leave single `copy` and `download` buttons that deliver the raw YAML.  That lets a user easily capture it and inspect it in their editor of choice, if needed.

jshaughn commented:
Oh, I'd leave the empty values, I prefer seeing the more complete picture.

ferhoyos commented:
I have converted tables from JSON to YAML. I had to upgrade `js-yaml` to the latest version to be able to add a `replacer` function in the `Additional State` tab to remove sensitive data like username. Additionally, I have removed the `json-beautify` library as there is no JSON parsing in Kiali now.

hhovsepy reviewed (approved):
YAML tables looks good:

![Screenshot from 2024-06-12 11-11-08](https://github.com/kiali/kiali/assets/604313/49b0ef87-55cf-4355-9ed1-0a3b2c98c848)
![Screenshot from 2024-06-12 11-10-47](https://github.com/kiali/kiali/assets/604313/87cf7824-b303-4381-b75a-ab68635b4226)

jshaughn reviewed (approved):
LG!
